File: python/ch20/parallel_sa.py
# python/ch20/parallel_sa.py
"""多開始 SA（シード並列）で Intro HC を改善（第20章）."""
from __future__ import annotations

import math
import random
import sys
import time
from concurrent.futures import ProcessPoolExecutor, as_completed
from dataclasses import dataclass
from typing import List, TextIO, Tuple

T0 = 5000.0
ALPHA = 0.995
TIME_LIMIT_SEC = 0.8
NUM_WORKERS = 4


@dataclass
class ProblemInput:
    days: int
    decay: List[int]
    gain: List[List[int]]


@dataclass
class ScheduleState:
    types: List[int]


def read_problem(stream: TextIO) -> ProblemInput:
    lines = stream.read().splitlines()
    idx = 0
    days = int(lines[idx])
    idx += 1
    decay = list(map(int, lines[idx].split()))
    idx += 1
    gain = [list(map(int, lines[idx].split())) for _ in range(days)]
    return ProblemInput(days=days, decay=decay, gain=gain)


def score_schedule(problem: ProblemInput, state: ScheduleState) -> int:
    last = [0] * 26
    satisfaction = 0
    for day in range(1, problem.days + 1):
        t = state.types[day - 1] - 1
        satisfaction += problem.gain[day - 1][t]
        last[t] = day
        decay = sum(problem.decay[i] * (day - last[i]) for i in range(26))
        satisfaction -= decay
    return max(10**6 + satisfaction, 0)


def random_neighbor(state: ScheduleState, rng: random.Random) -> ScheduleState:
    nxt = ScheduleState(list(state.types))
    day = rng.randrange(len(nxt.types))
    nxt.types[day] = rng.randint(1, 26)
    return nxt


def sa_run(
    problem: ProblemInput,
    seed: int,
    time_limit_sec: float,
) -> Tuple[List[int], int]:
    rng = random.Random(seed)
    state = ScheduleState([rng.randint(1, 26) for _ in range(problem.days)])
    score = score_schedule(problem, state)
    best_types = list(state.types)
    best_score = score
    temperature = T0
    start = time.perf_counter()
    step = 0

    while time.perf_counter() - start < time_limit_sec:
        nxt = random_neighbor(state, rng)
        new_score = score_schedule(problem, nxt)
        delta = new_score - score
        if delta >= 0 or rng.random() < math.exp(delta / temperature):
            state = nxt
            score = new_score
            if score > best_score:
                best_score = score
                best_types = list(state.types)
        temperature *= ALPHA
        step += 1

    print(f"worker seed={seed} best={best_score} steps={step}", file=sys.stderr)
    return best_types, best_score


def _worker(args: Tuple[ProblemInput, int, float]) -> Tuple[List[int], int]:
    problem, seed, limit = args
    return sa_run(problem, seed, limit)


def parallel_sa(problem: ProblemInput) -> Tuple[List[int], int]:
    seeds = [42, 43, 44, 45]
    results: List[Tuple[List[int], int]] = []

    try:
        with ProcessPoolExecutor(max_workers=NUM_WORKERS) as ex:
            futures = [
                ex.submit(_worker, (problem, seed, TIME_LIMIT_SEC)) for seed in seeds
            ]
            for fut in as_completed(futures):
                results.append(fut.result())
    except Exception as exc:
        print(f"parallel fallback ({exc})", file=sys.stderr)
        for seed in seeds:
            results.append(sa_run(problem, seed, TIME_LIMIT_SEC))

    best_types, best_score = max(results, key=lambda x: x[1])
    print(f"merged best score={best_score}", file=sys.stderr)
    return best_types, best_score


def main() -> None:
    problem = read_problem(sys.stdin)
    best_types, best_score = parallel_sa(problem)
    for t in best_types:
        print(t)
    print(f"output days={len(best_types)} score={best_score}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: go/ch20/parallel_sa.go
// go/ch20/parallel_sa.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	t0           = 5000.0
	alpha        = 0.995
	timeLimitSec = 0.8
	numWorkers   = 4
)

type ProblemInput struct {
	Days  int
	Decay []int
	Gain  [][]int
}

type ScheduleState struct {
	Types []int
}

type RunResult struct {
	Types []int
	Score int
	Seed  int
}

func readProblem(r *bufio.Reader) (ProblemInput, error) {
	line, err := r.ReadString('\n')
	if err != nil {
		return ProblemInput{}, err
	}
	days, _ := strconv.Atoi(strings.TrimSpace(line))
	line, err = r.ReadString('\n')
	if err != nil {
		return ProblemInput{}, err
	}
	decay := parseInts(strings.TrimSpace(line))
	gain := make([][]int, days)
	for i := 0; i < days; i++ {
		line, err = r.ReadString('\n')
		if err != nil {
			return ProblemInput{}, err
		}
		gain[i] = parseInts(strings.TrimSpace(line))
	}
	return ProblemInput{days, decay, gain}, nil
}

func parseInts(line string) []int {
	parts := strings.Fields(line)
	out := make([]int, len(parts))
	for i, p := range parts {
		out[i], _ = strconv.Atoi(p)
	}
	return out
}

func scoreSchedule(problem ProblemInput, types []int) int {
	last := make([]int, 26)
	satisfaction := 0
	for day := 1; day <= problem.Days; day++ {
		t := types[day-1] - 1
		satisfaction += problem.Gain[day-1][t]
		last[t] = day
		decay := 0
		for i := 0; i < 26; i++ {
			decay += problem.Decay[i] * (day - last[i])
		}
		satisfaction -= decay
	}
	score := 1000000 + satisfaction
	if score < 0 {
		return 0
	}
	return score
}

func randomNeighbor(types []int, rng *rand.Rand) []int {
	nxt := append([]int{}, types...)
	day := rng.Intn(len(nxt))
	nxt[day] = rng.Intn(26) + 1
	return nxt
}

func saRun(problem ProblemInput, seed int64, limit time.Duration) RunResult {
	rng := rand.New(rand.NewSource(seed))
	types := make([]int, problem.Days)
	for i := range types {
		types[i] = rng.Intn(26) + 1
	}
	score := scoreSchedule(problem, types)
	bestTypes := append([]int{}, types...)
	bestScore := score
	temperature := t0
	start := time.Now()
	step := 0

	for time.Since(start) < limit {
		nxt := randomNeighbor(types, rng)
		newScore := scoreSchedule(problem, nxt)
		delta := float64(newScore - score)
		if delta >= 0 || rng.Float64() < math.Exp(delta/temperature) {
			types = nxt
			score = newScore
			if score > bestScore {
				bestScore = score
				bestTypes = append([]int{}, types...)
			}
		}
		temperature *= alpha
		step++
	}

	fmt.Fprintf(os.Stderr, "worker seed=%d best=%d steps=%d\n", seed, bestScore, step)
	return RunResult{bestTypes, bestScore, int(seed)}
}

func parallelSA(problem ProblemInput) RunResult {
	seeds := []int64{42, 43, 44, 45}
	limit := time.Duration(timeLimitSec * float64(time.Second))
	results := make([]RunResult, len(seeds))
	var wg sync.WaitGroup
	wg.Add(len(seeds))
	for i, seed := range seeds {
		i, seed := i, seed
		go func() {
			defer wg.Done()
			results[i] = saRun(problem, seed, limit)
		}()
	}
	wg.Wait()

	best := results[0]
	for _, r := range results[1:] {
		if r.Score > best.Score {
			best = r
		}
	}
	fmt.Fprintf(os.Stderr, "merged best score=%d\n", best.Score)
	return best
}

func main() {
	problem, err := readProblem(bufio.NewReader(os.Stdin))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	best := parallelSA(problem)
	for _, t := range best.Types {
		fmt.Println(t)
	}
	fmt.Fprintf(os.Stderr, "output days=%d score=%d\n", len(best.Types), best.Score)
}
